sbom/main.ts

```diff
@@ -0,0 +1,131 @@
+        case 'pypi':
+            return `https://pypi.org/project/${pkg.version}/${pkg.version}/`;
+        case 'rubygems':
+            return `https://rubygems.org/gems/${pkg.name}/versions/${pkg.version}`;
+        case 'maven':
+            return `https://search.maven.org/artifact/${pkg.name}/${pkg.version}/pom`;
+        case 'nuget':
+            return `https://www.nuget.org/packages/${pkg.name}/${pkg.version}`;
+        default:
+            return 'NOASSERTION';
+    }
+}
+
+/**
+ *  Generates the document header for the produced SBOM.
+ */
+async function generateDocumentHeader(docNamespace: string) {
+    // TODO: need to capture
+    //  * DocumentNamespace (Not sure???)
+    //  * LicenseListVersion (Not sure???)
+    const docName = `${docNamespace}-SBOM`;
+    const today = new Date().toISOString();
+    const creatorComment = `SBOM Document for the ${docNamespace} project from Phylum, Inc.`;
+    return `##Document Header\n` + 
+           `SPDXVersion: ${spdxVersion}\n` +
+           `DataLicense: ${dataLicense}\n` +
+           `SPDXID: SPDXRef-Document\n` +
+           `DocumentName: ${docName}\n` +
+           `DocumentNamespace: ${docNamespace}\n` + 
+           `Creator: ${creator}\n` +
+           `Created: ${today}\n` +
+           `CreatorComment: <text>${creatorComment}</text>\n\n`;
+}
+
+/**
+ *  Creates a package block for the provided package JSON.
+ */
+function createPackageBlock(pkg: Package) {
+    return `##### Package: ${pkg.name}\n\n` +
+           `PackageName: ${pkg.name}\n` +
+           `SPDXID: SPDXRef-${pkg.name}-${pkg.version}\n` +
+           `PackageVersion: ${pkg.version}\n` +
+           `PackageDownloadLocation: NOASSERTION\n` + // TODO: Get this!
+           `FilesAnalyzed: true\n` +
+           `PackageHomePage: ${getHomepage(pkg)}\n` +
+           `PackageLicenseConcluded: NOASSERTION\n` + // TODO: Get this!
+           `PackageLicenseDeclared: ${pkg.license}\n` +
+           `PackageCopyrightText: NOASSERTION\n` +
+           `ExternalRef: PACKAGE-MANAGER ${pkg.type} ${pkg.name}@${pkg.version}\n\n`
+}
+
+/**
+ *  Creates a relationship entry for the provided package JSON.
+ */
+function createRelationship(pkg: string, ver: string, dep: Dependency) {
+    return `Relationship: SPDXRef-${dep.name}-${dep.version} PREREQUISITE_FOR SPDXRef-${pkg}-${ver}\n`;
+}
+
+/**
+ *  Given a Phylum job JSON, generates the SBOM document and drops
+ *  it to disk.
+ */
+async function toSBOM(data: string) {
+    let projectName = data["project_name"];
+    let sbom = await generateDocumentHeader(projectName);
+    let refs = "";
+
+    // Construct the packages section of our SBOM
+    data["packages"].forEach((p) => {
+        let pkg: Package = p;
+        sbom += createPackageBlock(pkg);
+
+        let pkgName: keyof typeof obj;
+
+        // Create the package relationships of our SBOM
+        for (pkgName in pkg.dependencies) {
```

Comment:
This works because `for... in` iterates through array indices: https://www.typescriptlang.org/docs/handbook/iterators-and-generators.html#forof-vs-forin-statements

Given that `pkg.dependencies` is an array of `Dependency`, though, we could use `pkg.dependencies.forEach` or `for... of` and immediately have `dep` instead of having to construct `dep` 🤷‍♂️ 

But this works.